Add bubble elements with application to obstacle problem

pazner opened the pull request:
This PR adds conforming bubble finite elements, where the shape functions are given by $\phi(\boldsymbol{x}) + b(\boldsymbol{x}) \psi(\boldsymbol{x})$, where $\phi \in \mathcal{P}_p$, $\psi \in \mathcal{P}_q$, and $b$ is the lowest-order bubble function.

The use of the element is illustrated in the obstacle problem in ex36 and ex36p.
<!--GHEX{"author":"pazner","assignment":"2025-12-16T20:46:57","editor":"tzanio","id":5140,"reviewers":["sohailreddy","tzanio","jdongg"],"merge":"2026-01-06T20:46:57","approval":"2025-12-30T20:46:57"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#5140](https://github.com/mfem/mfem/pull/5140) | @pazner | @tzanio | @sohailreddy + @tzanio + @jdongg | 12/16/25 | ⌛due 12/30/25 | ⌛due 1/6/26 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

najlkin committed:
Added extrusion of vector 1D grid functions.

mlstowell committed:
Adding new MapType entries for R2D and R1D classes

mlstowell committed:
Using new MapType entries and implementing new GetPhys*Dim methods

mlstowell committed:
Updating unit tests

Sbozzolo committed:
Improve error message for gmsh versions != 2.2

I am a new user of [palace](https://github.com/awslabs/palace). As I was
trying to set a simple mesh up (with gmsh), I kept getting indexing
errors that I could not decipher. I eventually
[learned](https://mfem.org/mesh-formats/) that supported version for
gmsh meshes is 2.2.

This commit catches this and adds an informative error.

thartland committed:
adding a check to make sure that each process owns at least one entry of the HypreParVector prior to calling GlobalVector

thartland committed:
style

thartland committed:
VERIFY instead of ASSERT

najlkin committed:
Added documentation and checks to the extrusion classes.

helloworld922 committed:
improving config.mk file generated by cmake to work with hip/cuda

Still need to export compiler flags

helloworld922 committed:
seems to be building external laghos now

helloworld922 committed:
remove debug printout

helloworld922 committed:
fixes for hip

pazner committed:
Add H1 bubble triangle element and collection

pazner committed:
Add quad bubble element, change meaning of q

pazner committed:
Move bubble elements to their own file

pazner committed:
Disable check that FE and FEC orders are the same

With enriched bubble elements, the orders could be different.

For example, linear triangle enriched with bubble has max total degree 3, but
the linear quadrilateral enriched with bubble has max degree 2 in each variable
(and max total degree 4).

pazner committed:
Add bubble tets and hexes

pazner committed:
Return nullptr for H1Bubble_FECollection::DofOrderForOrientation

Some features (e.g. node reordering) won't be supported; this could be added
later.

helloworld922 committed:
review suggestions

helloworld922 committed:
missed one old unsetting of shared_link_flag

pazner committed:
Revert "Return nullptr for H1Bubble_FECollection::DofOrderForOrientation"

This reverts commit 58e7bb6e6e213eb90839feb67de7d0e03b5799da.

pazner committed:
Fix DOF orderings in bubble elements

pazner committed:
Use bubble elements in ex36 and ex36p

dohyun-cse committed:
FEColl::New

pazner committed:
Support "H1Bubble@" in FiniteElementCollection::New

pazner committed:
Improve Doxygen

pazner committed:
Re-add assertion; skip check for bubble spaces

pazner committed:
Small fixes

Add local variables in thread-safe mode
Fix MFEM_VERIFY message
Fix trace collection order

pazner commented on fem/fe/fe_h1_bubble.cpp:
```diff
@@ -0,0 +1,972 @@
+#ifdef MFEM_THREAD_SAFE
+   const int n1d = max(p + 1, q + 1);
+   const int npq = (p+1)*(p+1)*(p+1) + (q+1)*(q+1)*(q+1);
+   Vector shape_x(n1d), shape_y(n1d), dshape_x(n1d), dshape_y(n1d);
```
Resolved in 456c236cc522b1963c1586beb4864904a3b14fa8

pazner commented on fem/fe_coll.cpp:
```diff
@@ -2122,6 +2132,172 @@ H1_FECollection::~H1_FECollection()
+     bubble_order(q)
+{
+   MFEM_VERIFY(p >= 1, "H1Bubble_FECollection requires order >= 1.");
+   MFEM_VERIFY(dim >= 0 && dim <= 3, "H1_FECollection requires 0 <= dim <= 3.");
```
Resolved in 456c236cc522b1963c1586beb4864904a3b14fa8.

brendankeith reviewed (approved):
This looks good to me. Computing the orthogonal projection at the element level is clever, although perhaps not optimally efficient.

I think this is good to go once we get one more review.

dohyun-cse committed:
Merge branch 'master' into bubble

najlkin committed:
Fixed visulization in ex22p.

helloworld922 committed:
suggestions from Veselin

najlkin committed:
Fixed complex grid function copy assignment.

najlkin committed:
Minor docstring correction.

najlkin committed:
Fixed vis of the initial exact solution.

helloworld922 committed:
revert change, updated comment to why libdl gets special treatment

najlkin committed:
Removed unused code.

najlkin committed:
Generalized RT normal projection.

najlkin committed:
Minor unification of docstrings.

mlstowell committed:
Adding plasma miniapp directory

mlstowell committed:
Adding miniapps/plasma subdirectory to build system

pazner committed:
Fix bug in ParL2FaceRestriction with vdim > 1

The layout of the FaceNbrData vector was not handled properly

pazner committed:
Add unit test for parallel L2 face restriction with vdim > 1

pazner committed:
Add comment about the shape of FaceNbrData

balay committed:
rework PetscContainerSetCtxDestroy() usage for < (3,23,0)

balay committed:
update PetscCtxDestroyFn usage for < (3,23,0)

balay committed:
update KSPMonitorFn usage for < (3,24,0)

BarrySmith committed:
Update to use PetscCtxRt from (3,25,0), and cleanup duplicate code

Co-authored-by: Nuno Nobre <nuno.nobre@stfc.ac.uk>
Co-authored-by: Satish Balay <balay@mcs.anl.gov>

chapman39 committed:
dfem integrate: use mfem abort kernel in device code

chapman39 committed:
eliminate usage of modulus to avoid llvm backend bug

chapman39 committed:
80 chars/ line

Commit:
added missing face orientation

helloworld922 committed:
MFEM_EXPORT_GPU_CONFIG should export CPU config.mk when set to off

hughcars committed:
Add positive-weight simplex quadrature rules for orders 0-20

Triangle rules from Witherden & Vincent (2015), tet rules d=0-13
from Witherden & Vincent, tet rules d=14-20 from Chuluunbaatar et al.
(2022). All rules have strictly positive weights and interior points,
replacing the legacy rules which use negative weights at several
orders and fall back to Grundmann-Moller (negative weights, high
point counts) for tets at d>=9.

hughcars committed:
Fix memory leak

helloworld922 committed:
Update config/cmake/modules/MfemCmakeUtilities.cmake

Co-authored-by: Nuno Nobre <nuno.nobre@stfc.ac.uk>

helloworld922 committed:
fixed wrong dir being marked as system

helloworld922 committed:
Update config/cmake/modules/MfemCmakeUtilities.cmake

Co-authored-by: Nuno Nobre <nuno.nobre@stfc.ac.uk>

helloworld922 committed:
Update config/cmake/modules/MfemCmakeUtilities.cmake

Co-authored-by: Nuno Nobre <nuno.nobre@stfc.ac.uk>

pazner committed:
Merge remote-tracking branch 'origin/master' into bubble

# Conflicts:
#	fem/fe_coll.hpp

mlstowell committed:
Changing copyright date to pass CI checks

helloworld922 committed:
move cudart to MFEM_EXT_LIBS

najlkin committed:
Made the ProjectBdrCoefficientNormal check non-debug.

helloworld922 committed:
comment on why TPL_LIBRARIES is reversed twice

najlkin committed:
Added a unit test for vector ProjectBdrCoefficientNormal().

najlkin committed:
Added scalar unit test of ProjectBdrCoefficientNormal().

jdongg committed:
fix clang compiler warnings from origin/catch-tests

chapman39 committed:
added comment

Commit:
extended to MixedBilinearForm

hughcars committed:
Add existing order 21-25 triangle rule to positive-weight rules

The 126-point degree-25 rule already has all positive weights.
Copy it into TrianglePositiveIntegrationRule so the positive-weight
path covers orders 0-25.

chapman39 committed:
add comments showing each modulus replacement

hughcars committed:
Remove legacy simplex rules; positive-weight rules are now the default

The positive-weight rules now cover the full tabulated range for both
triangles (0-25) and tetrahedra (0-20), so the old rules with negative
weights are no longer needed. Remove the SimplexQuadrature enum,
simplex_type member, and legacy rule functions — all simplex quadrature
now uses positive-weight rules by default, with Grundmann-Moller
fallback for higher orders.

hughcars committed:
Restore original function order in intrules.cpp

Move TriangleIntegrationRule before SquareIntegrationRule to match
the original file layout, reducing diff noise against master.

hughcars committed:
Use exact fractions for trivial quadrature weights and coordinates

For rules where the mathematical value is an exact simple fraction
(midpoint weights, equal-weight symmetric rules), use the fraction
directly rather than the Polyquad decimal expansion. Cleaner to read
and avoids any rounding from decimal-to-double conversion.

hughcars committed:
Remove unused private helper methods from IntegrationRule

AddTriPoints3R, AddTetPoints4b, and AddTetPoints12bc are no longer
called after the legacy simplex rules were removed.

pazner committed:
Use constexpr in unit test

v-dobrev committed:
Small change in error message + formatting.

mlstowell committed:
Adding bugfix and unit test which would have caught the bug

najlkin committed:
Fixed spelling of transverse.

hughcars committed:
Address PR feedback

- Use [IntegrationRules] test tag instead of [PositiveWeightRules]
- Remove redundant case 21: (default branch handles it via the overwrite guard)
- Remove trailing blank line

pazner committed:
Fix signed char issue in Device::GetUUID

v-dobrev committed:
Update actions/{checkout,cache} to v5

Update github/codeql-action/* to v3

v-dobrev committed:
Updated actions/checkout to the latest major version, v6

v-dobrev committed:
Updated the github/codeql-action/* actions to the latest, v4

adam-sim-dev committed:
Fixed missing parenthesis in the comment

v-dobrev committed:
Update the action actions/cache/restore to v5

v-dobrev committed:
Adjust seed values in sample runs in ex12p to ensure LOBPCG convergence in
older hypre versions.
